### PR TITLE
[AIRFLOW-1467] Dynamic pooling via allowing tasks to use more than one pool slot (depending upon the need)

### PR DESCRIPTION
**Pull request by lokeshlal**
PR contains changes in pool and task instance to provide functionality to tasks to use more than one pool slot.

- Added pool_capacity field in TaskInstance (pool_capacity is defaulted to 1, to maintain the current behavior)
- Added pool_capacity in baseoperator
- Modified pools functionality to calculate the used/queued/occupied/available slots
- Modified pool_slots_available_dep.py to check against task _pool_capacity_ field instead of 1 
- Modifed test case in file test_pool_slots_available_dep.py to include pool_capacity in the Mock object

-----------------------------------------------------------------------------------------------

- [x] Description above provides context of the change
- [X] Commit message contains [\[AIRFLOW-1467\]](https://issues.apache.org/jira/browse/AIRFLOW-1467) or `[AIRFLOW-XXXX]` for document-only changes
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by tooptoop4** [comment]
https://issues.apache.org/jira/browse/AIRFLOW-6227 could be another approach

**Comment by lokeshlal** [comment]
@tooptoop4 Yes the approach looks good when multiple pools are required as described in the jira ticket.  
This PR will be useful in a scenario, where we have a spark cluster where jobs needs to be submitted with different complexity (such as Large jobs, medium jobs etc) and each job would require different capacity on spark cluster. Hence dynamic pooling can help control the spark cluster capacity directly from the Airflow using pools. this is aligned to the following jira ticket https://issues.apache.org/jira/browse/AIRFLOW-1467 

The problem statement mentioned in the jira ticket https://issues.apache.org/jira/browse/AIRFLOW-6227, can be handled via locking a file for write. That is, if the ask is to keep one writer on a table, then before triggering spark job, create another task that will put a file write lock on a file (name same as table name) in the file system (libraries such as fasteners or lockfile in a python operator can be used). This will make sure that at a time only one job will be triggered for the said table and makes the code more dynamic rather than creating pools every time a new table is introduced. and once the spark job finishes (weather the job fail or pass) then release the lock from the file.

**Comment by dimberman** [comment]
@lokeshlal +1 this is a great idea!

Could you please make the unit tests a bit more comprehensive? (e.g. multiple pools)

**Comment by lokeshlal** [comment]
@dimberman Sorry for the confusion I created with my previous comment. I have not implemented multiple pools... only implemented a way to use more than one pool slot in a task. specifically trying to resolve https://issues.apache.org/jira/browse/AIRFLOW-1467 

for example, provide number of pool slots a task will be using using pool_capacity property
```
task = BashOperator(
    task_id='sleep',
    bash_command='sleep 300',
    retries=3,
    pool='test',
    pool_capacity=6,	
    dag=dag)
```

**Comment by lokeshlal** [comment]
@dimberman I have updated pools test cases for task instances. Could you please review the code again. Thank you.

File: airflow/models/taskinstance.py
Comment (xinbinhuang) [inline airflow/models/taskinstance.py]: Since the attribute name is the same as the `Column.name`, I suggest removing the `Column.name` and use the default mapping to be consistent with the rest.

**Review by xinbinhuang — CHANGES_REQUESTED** [review]

File: airflow/models/baseoperator.py
Comment (xinbinhuang) [inline airflow/models/baseoperator.py:315]: Should `pool_capacity` be ensured to `>= 1`? Otherwise, we may possibly end up with negative used/queued slots, which will make the calculated `open_slots` > preset `slots`  (see [here](https://github.com/apache/airflow/blob/ed221501ae51d5b0d7b4d60bc928d10e4f971a02/airflow/models/pool.py#L113) ). Any thoughts? @dimberman @ashb @mik-laj

Comment (lokeshlal) [reply airflow/models/baseoperator.py:315]: Yes, it is correct. pool_capacity should never be less than 1. I will put a check in taskinstance __init__ for pool_capacity value to be >= 1.

File: airflow/models/taskinstance.py
Comment (lokeshlal) [reply airflow/models/taskinstance.py]: I have modified the code as suggested.

File: airflow/models/baseoperator.py
Comment (lokeshlal) [reply airflow/models/baseoperator.py:315]: I have added a check in taskinstance and also added a corresponding test case against the same.

**Comment by dimberman** [comment]
@lokeshlal why do all of the test cases have priority_weight of 1? Could you do a test of weight=2 and of weight=0?

**Comment by lokeshlal** [comment]
@dimberman  Yes I have modified test cases to include pool_capacity to be 0, 1 and 2.  I have again modified to include the pool_capacity in other test cases as well. the build is under the process. hopefully I have not missed the any other test case.

**Review by dimberman — APPROVED** [review]

**Comment by dimberman** [comment]
Thank you @lokeshlal. I'll merge this once tests pass.

**Comment by lokeshlal** [comment]
Thank you @dimberman.

File: airflow/models/baseoperator.py
Comment (xinbinhuang) [reply airflow/models/baseoperator.py:315]: Looks good! Thanks @lokeshlal

**Comment by lokeshlal** [comment]
Hi @dimberman, I have done changes in few more test cases to pass pool_capacity property in the Mock object. Test cases also executed successfully. Could you please review the latest commit as well (commit after your approval).

**Comment by lokeshlal** [comment]
Hi @dimberman, Could you please review the remaining changes as well and if everything looks good, please merge the changes as well. thanks.

**Comment by dimberman** [comment]
@lokeshlal LGTM thanks for setting this up!

**Comment by potiuk** [comment]
Hello @lokeshlal @dimberman - I had to revert this change from master as it created two  heads in sqlalchemy migrations and we have test to detect that. The root cause was that this PR was not rebased after the second head was added (otherwise it would have been detected by failing Travis CI run - last time Travis was run 6 days ago for this PR and the other head (#6489 ) was merged 3 days ago. 

It's no-one fault, it can happen in our process, but we will have to add better protection against such problems in the future. In the meantime can you please create another PR with this change, fix the duplicated heads (by making your migration based on the new head migration) and re-submit/merge the PR again.

Sorry for the inconvenience @lokeshlal - I understand it was your first PR merged and you hit this - very rare - problem :(. We will do better in the future to avoid it.

**Comment by potiuk** [comment]
Just in case: https://blog.jerrycodes.com/multiple-heads-in-alembic-migrations/

**Comment by potiuk** [comment]
I prefer the "hacky" way.

**Comment by lokeshlal** [comment]
Thank you @dimberman and @potiuk - I have created another PR with a single migration head. https://github.com/apache/airflow/pull/7160
It is in travis buid. hopefully test will pass.

**Comment by potiuk** [comment]
We still have "kerberos" related instability. I restarted the build and hopefully it will pass.

**Comment by potiuk** [comment]
(working to fix the instability as well)

**Pull request merged**
